// nocust-client/dist/wallet/checks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = require("bignumber.js");
var activeState_1 = require("./activeState");
var blockchain_1 = require("../services/blockchain");
var store_1 = require("../store");
var constants_1 = require("../constants");
exports.checkRegistration = function (address, token, registration) {
    var clientRegistrationHash = activeState_1.hashActiveState({
        token: token,
        address: address,
        eon: registration.eon,
        trailIdentifier: 0,
        transactionSetHash: constants_1.EMPTY_HASH,
        spent: new bignumber_js_1.BigNumber(0),
        gained: new bignumber_js_1.BigNumber(0),
    });
    var isClientRegistrationValid = blockchain_1.blockchain.validateSignature(clientRegistrationHash, registration.walletSignature, address);
    if (!isClientRegistrationValid)
        return false;
    var operatorRegistrationHash = activeState_1.hashActiveState({
        address: address,
        token: token,
        eon: registration.eon,
        trailIdentifier: registration.trailIdentifier,
        transactionSetHash: constants_1.EMPTY_HASH,
        spent: new bignumber_js_1.BigNumber(0),
        gained: new bignumber_js_1.BigNumber(0),
    });
    var isOperatorRegistrationValid = blockchain_1.blockchain.validateSignature(operatorRegistrationHash, registration.operatorSignature, store_1.store.contractOwner);
    if (!isOperatorRegistrationValid)
        return false;
    return true;
};
