// nocust-client/dist/wallet/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var store_1 = require("../store");
var utils_1 = require("../helpers/utils");
var storage_1 = require("../services/storage");
var signature_1 = require("./signature");
var eon_1 = require("./eon");
exports.getWallet = function (address, token) {
    return store_1.store.wallets.get(utils_1.defineWalletKey(address, token));
};
function updateWallet(wallet) {
    store_1.store.wallets.set(utils_1.defineWalletKey(wallet.address, wallet.token), wallet);
    storage_1.storage.queueWalletStorage(wallet);
}
exports.updateWallet = updateWallet;
exports.createWallet = function (address, token, parentAddress) { return ({
    address: address,
    token: token,
    isRegistered: false,
    parentAddress: parentAddress,
}); };
function fillWalletRegistration(wallet, data) {
    if (wallet.isRegistered)
        throw new Error('[INTERNAL] Wallet is already registered!');
    wallet.registrationSignature = signature_1.createSignature(wallet.address, data.hash, data.walletSignature);
    wallet.operatorRegistrationSignature = signature_1.createSignature(store_1.store.contractOwner, data.hash, data.operatorSignature);
    wallet.isRegistered = true;
    wallet.registrationEon = data.eon;
    wallet.trailIdentifier = data.trailIdentifier;
    wallet.currentEon = new eon_1.Eon(wallet, data.eon);
}
exports.fillWalletRegistration = fillWalletRegistration;
