// nocust-client/dist/helpers/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var NCErrorCode;
(function (NCErrorCode) {
    NCErrorCode["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
    NCErrorCode["INITIALIZATION_FAILED"] = "INITIALIZATION_FAILED";
    NCErrorCode["WALLET_UNREGISTERED"] = "WALLET_UNREGISTERED";
    NCErrorCode["RECIPIENT_UNREGISTERED"] = "RECIPIENT_UNREGISTERED";
    NCErrorCode["WALLET_ALREADY_REGISTERED"] = "WALLET_ALREADY_REGISTERED";
    NCErrorCode["CAN_NOT_RECOVER_FUNDS"] = "CAN_NOT_RECOVER_FUNDS";
    NCErrorCode["NO_PRIVATE_KEY"] = "NO_PRIVATE_KEY";
    NCErrorCode["SLA_NOT_EXPIRED_YET"] = "SLA_NOT_EXPIRED_YET";
    NCErrorCode["PREPARE_TRANSFER_FAILURE"] = "PREPARE_TRANSFER_FAILURE";
    NCErrorCode["POST_FAILURE"] = "POST_FAILURE";
    NCErrorCode["REGISTRATION_THROTTLING"] = "REGISTRATION_THROTTLING";
    NCErrorCode["FETCH_OPERATOR_DATA_ERROR"] = "FETCH_OPERATOR_DATA_ERROR";
    NCErrorCode["FETCH_PARENT_CHAIN_DATA_ERROR"] = "FETCH_PARENT_CHAIN_DATA_ERROR";
    NCErrorCode["CONFIRMATION_TIMEOUT"] = "CONFIRMATION_TIMEOUT";
    NCErrorCode["NON_UNIQUE_TRANSFER"] = "NON_UNIQUE_TRANSFER";
    NCErrorCode["TX_HASH_NOT_FOUND"] = "TX_HASH_NOT_FOUND";
    NCErrorCode["PARENT_CHAIN_TRANSACTION_FAILURE"] = "PARENT_CHAIN_TRANSACTION_FAILURE";
    NCErrorCode["INSUFFICIENT_WITHDRAWAL_LIMIT"] = "INSUFFICIENT_WITHDRAWAL_LIMIT";
    NCErrorCode["NO_ETHER_TO_PAY_FOR_GAS"] = "NO_ETHER_TO_PAY_FOR_GAS";
    NCErrorCode["SUB_ACCOUNT_INDEX_NOT_FOUND"] = "SUB_ACCOUNT_INDEX_NOT_FOUND";
    NCErrorCode["SWAP_LIMIT_REACHED_ERROR"] = "SWAP_LIMIT_REACHED_ERROR";
    NCErrorCode["INSUFFICIENT_COMMIT_CHAIN_BALANCE"] = "INSUFFICIENT_COMMIT_CHAIN_BALANCE";
    NCErrorCode["INSUFFICIENT_PARENT_CHAIN_BALANCE"] = "INSUFFICIENT_PARENT_CHAIN_BALANCE";
    NCErrorCode["INSUFFICIENT_TRANSFER_ALLOWANCE"] = "INSUFFICIENT_TRANSFER_ALLOWANCE";
    NCErrorCode["SWAP_PENDING"] = "SWAP_PENDING";
    NCErrorCode["TRANSFER_NOT_FOUND"] = "TRANSFER_NOT_FOUND";
    NCErrorCode["TRANSFER_TO_SELF"] = "TRANSFER_TO_SELF";
    NCErrorCode["NO_TRADING_SETUP"] = "NO_TRADING_SETUP";
    NCErrorCode["CAN_NOT_CANCEL_SWAP"] = "CAN_NOT_CANCEL_SWAP";
    NCErrorCode["CAN_NOT_ISSUE_STATE_UPDATE_CHALLENGE"] = "CAN_NOT_ISSUE_STATE_UPDATE_CHALLENGE";
    NCErrorCode["CAN_NOT_ISSUE_DELIVERY_CHALLENGE"] = "CAN_NOT_ISSUE_DELIVERY_CHALLENGE";
})(NCErrorCode = exports.NCErrorCode || (exports.NCErrorCode = {}));
var NCServerErrorCode;
(function (NCServerErrorCode) {
    NCServerErrorCode["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
    NCServerErrorCode["EON_NUMBER_OUT_OF_SYNC"] = "EON_NUMBER_OUT_OF_SYNC";
    NCServerErrorCode["NOT_FOUND"] = "NOT_FOUND";
    NCServerErrorCode["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    NCServerErrorCode["NON_UNIQUE"] = "unique";
})(NCServerErrorCode = exports.NCServerErrorCode || (exports.NCServerErrorCode = {}));
var NCError = (function (_super) {
    __extends(NCError, _super);
    function NCError(code, description, data) {
        var _this = _super.call(this, code) || this;
        _this.description = description;
        _this.data = data;
        return _this;
    }
    return NCError;
}(Error));
exports.NCError = NCError;
var NCServerError = (function (_super) {
    __extends(NCServerError, _super);
    function NCServerError(code, description, data) {
        var _this = _super.call(this, code) || this;
        _this.description = description;
        _this.data = data;
        return _this;
    }
    return NCServerError;
}(Error));
exports.NCServerError = NCServerError;
