// nocust-client/dist/helpers/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = __importDefault(require("bignumber.js"));
exports.remove0x = function (address) {
    return address.indexOf('0x') === 0 ? address.substring(2) : address;
};
exports.defineWalletKey = function (address, token) { return token + "/" + address; };
exports.failOnTimeout = function (timeoutSeconds) {
    return new Promise(function (resolve, reject) { return setTimeout(reject, timeoutSeconds * 1000); });
};
exports.isSameHexValue = function (a, b) { return a.toLowerCase() === b.toLowerCase(); };
exports.generateRandomNonce = function () {
    return new bignumber_js_1.default(Math.floor(100000000 + Math.random() * 899999999));
};
exports.sleep = function (ms) { return new Promise(function (resolve) { return setTimeout(resolve, ms); }); };
function detectEnvironment() {
    if (typeof document !== 'undefined') {
        return 'browser';
    }
    else if (typeof navigator !== 'undefined' && navigator.product === 'ReactNative') {
        return 'react-native';
    }
    else {
        return 'node';
    }
}
exports.detectEnvironment = detectEnvironment;
