// nocust-client/dist/nocust/withdraw.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("./index");
var errors_1 = require("../helpers/errors");
var contracts_1 = require("../services/contracts");
var blockchain_1 = require("../services/blockchain");
var operator_1 = require("../services/operator");
function withdraw(config) {
    return __awaiter(this, void 0, void 0, function () {
        var address, token, amount, gasPrice, limit, e_1, lastSubmittedEon, walletState, tx, e_2, hashPosition, txId, _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    address = config.address, token = config.token, amount = config.amount, gasPrice = config.gasPrice;
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 3, , 4]);
                    return [4, index_1.nocust.getWithdrawalLimit(address, token)];
                case 2:
                    limit = _d.sent();
                    if (limit.isLessThan(amount)) {
                        return [2, Promise.reject(new errors_1.NCError(errors_1.NCErrorCode.INSUFFICIENT_WITHDRAWAL_LIMIT))];
                    }
                    return [3, 4];
                case 3:
                    e_1 = _d.sent();
                    return [2, Promise.reject(e_1)];
                case 4:
                    _d.trys.push([4, 8, , 11]);
                    return [4, index_1.nocust.getLastSubmittedEon()];
                case 5:
                    lastSubmittedEon = _d.sent();
                    return [4, operator_1.operator.fetchWalletState({
                            address: address,
                            token: token,
                            eon: lastSubmittedEon - 1,
                        })];
                case 6:
                    walletState = _d.sent();
                    return [4, contracts_1.contracts.requestWithdrawal({
                            address: address,
                            amount: amount,
                            gasPrice: gasPrice,
                            token: token,
                            walletState: walletState,
                        })];
                case 7:
                    tx = _d.sent();
                    return [2, tx];
                case 8:
                    e_2 = _d.sent();
                    console.log('error', e_2);
                    hashPosition = e_2.message.indexOf('transactionHash');
                    if (!(hashPosition > -1)) return [3, 10];
                    txId = e_2.message.substr(hashPosition + 19, 66);
                    console.log('txId', txId);
                    _b = (_a = console).log;
                    _c = ['EVM revert reason:'];
                    return [4, blockchain_1.blockchain.fetchEVMRevertMessage(txId)];
                case 9:
                    _b.apply(_a, _c.concat([_d.sent()]));
                    _d.label = 10;
                case 10: return [2, Promise.reject(e_2)];
                case 11: return [2];
            }
        });
    });
}
exports.default = withdraw;
